plugins/modules/source_control/gitlab/gitlab_integration.py

```diff
@@ -0,0 +1,544 @@
+            if not self._module.check_mode:
+                remote.save()
+            return remote
+
+        return False
+
+    def __as_api_object(self, active, params, events):
+        obj = {'active': active}
+        obj.update(self.__expand_events(events))
+        obj.update(params)
+        return obj
+
+    def __expand_events(self, events):
+        try:
+            # See https://gitlab.com/gitlab-org/gitlab/-/issues/27050 for why it's useful to
+            # discard unsupported events before comparing
+            supported_events = SRV_DEF[self.name]['_events']
+        except KeyError:
+            supported_events = None
+        ret = dict((value + '_events', bool(value in events)) for value in self.HOOK_EVENTS if supported_events is None or value in supported_events)
+        return ret
+
+    @staticmethod
+    def __events_equal(obj, events):
+        return all((k in obj and obj[k] == v) for k, v in events.items())
+
+    @staticmethod
+    def __has_credentials(attr):
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def __equals(self, prev_attr, active, params, events):
+        filtered_params = dict((k, v) for k, v in params.items() if k not in self.CREDENTIAL_PARAMS)
+        if prev_attr['active'] != active:
+            self._module.debug("active status differs")
+            return False
+        if self.__has_credentials(params):
+            self._module.debug("has credentials: assuming difference")
+            return False
+        if not self.__events_equal(prev_attr, self.__expand_events(events)):
+            self._module.debug("events differs: %s != %s" % (prev_attr, events))
+            return False
+        if prev_attr['properties'] != filtered_params:
+            self._module.debug("integrations attributes differs: %s != %s" % (prev_attr['properties'], filtered_params))
+            return False
+        return True
+
+    def setattrs(self, obj, new_attr):
+        for k, v in new_attr.items():
+            setattr(obj, k, v)
+
+
+# Equivalent of doing `AnsibleModule(argument_spec=base_spec, **constraints)`
+def mimic_ansible_module_validation(specs, constraints):
+    from ansible.module_utils.basic import ModuleArgumentSpecValidator, _load_params
+    from ansible.module_utils.errors import UnsupportedError
+    import os
+
+    params = _load_params()
+    first_level_validator = ModuleArgumentSpecValidator(specs, **constraints)
+    validation_result = first_level_validator.validate(params)
+    try:
+        error = validation_result.errors[0]
+    except IndexError:
+        error = None
+    # Fail for validation errors, even in check mode
+    if error:
+        msg = validation_result.errors.msg
+        if isinstance(error, UnsupportedError):
+            msg = "Unsupported parameters for ({name}) {kind}: {msg}".format(name=os.path.basename(__file__), kind='module', msg=msg)
```

Comment:
There is no need to use `os.path.basename()` here.

Reply:
Why?
(`lib/ansible/module_utils/basic.py` does it that way and I stuck to the standard error output)

Reply:
Because this is a specific module, not a module utils.